### ISSUE TITLE
cython: compilation error while building h5py (solved by workaround)

### ISSUE THREAD

**Issue by d-a-v** [h5py/h5py#1533]
Hello, here's the error-logs and steps to reproduce when I try to rebuild h5py with python3/pip on ubuntu18.04.

 * Operating System Ubuntu 18.04
 * Python version 3.6
 * Where Python was acquired: system Python on Linux
 * h5py version: 2.10.0 (from pip)
 * HDF5 version: 1.10.0

I provide a docker file to reproduce the problem locally, that you can run from a clean directory with
```
$ mv Dockerfile.txt Dockerfile
$ docker build .
```

Logs: 
```
[...]
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
    
    
    # --- open, open_by_name, open_by_idx ---
    
    @with_phil
    def open(ObjectID loc not None, char* name=NULL, int index=-1, *,
                                   ^
    ------------------------------------------------------------
    
    h5py/h5a.pyx:54:32: Cannot convert 'void *' to Python object
[...]
```

[Dockerfile.txt](https://github.com/h5py/h5py/files/4482778/Dockerfile.txt)

which essentially does:
```
RUN python3 -m pip install --no-cache-dir mpi4py
RUN echo ------- available system libraries:; find /usr/lib -name "libhdf5.so*"; echo -------
RUN LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu/hdf5/serial  python3 -m pip install --no-cache-dir --no-binary=h5py h5py
```

Full log: [log.txt](https://github.com/h5py/h5py/files/4482817/log.txt)

**Comment by ululi1970** [comment]
I am having exactly the same issue. This is on a clean ubuntu 18.04 install. 
Odd thing is, until last week I had not problem compiling it with MPI support.

**Comment by ululi1970** [comment]
Compiling the serial version fails with the same problem 
HDF5_DIR=/usr/lib/x86_64-linux-gnu/hdf5/serial/ pip3 install --no-binary=h5py h5py --user

...
 Error compiling Cython file:
    ------------------------------------------------------------
    ...


    # --- open, open_by_name, open_by_idx ---

    @with_phil
    def open(ObjectID loc not None, char* name=NULL, int index=-1, *,
                                   ^
    ------------------------------------------------------------

    h5py/h5a.pyx:54:32: Cannot convert 'void *' to Python object

    Error compiling Cython file:
    ------------------------------------------------------------
    ...
        """
        H5Arename_by_name(loc.id, obj_name, name, new_name, pdefault(lapl))


    @with_phil
    def delete(ObjectID loc not None, char* name=NULL, int index=-1, *,
                                     ^
    ------------------------------------------------------------

    h5py/h5a.pyx:122:34: Cannot convert 'void *' to Python object

    Error compiling Cython file:
    ------------------------------------------------------------
    ...
        def _hash(self):
            return hash((self.corder_valid, self.corder, self.cset, self.data_size))


    @with_phil
    def get_info(ObjectID loc not None, char* name=NULL, int index=-1, *,
                                       ^
    ------------------------------------------------------------
 h5py/h5a.pyx:184:36: Cannot convert 'void *' to Python object

**Comment by d-a-v** [comment]
> Odd thing is, until last week I had not problem compiling it with MPI support.

It indeed seems to have appeared some days ago.

**Comment by ululi1970** [comment]
It looks like there has been a recent change to the Cython API which must have broken h5py as fetched by pip.  
Cloning the master branch on my machine and compiling from it worked all right. Hopefully, the fixed code will make its way to the tarball that pip fetches from.

**Comment by Smankusors** [comment]
this happens to me too

I think this is only happen when there's no compiled build/wheel available, and using Cython to build it.

**Comment by ululi1970** [comment]
When using Cython, it seems necessary to use the latest code from the master branch.

**Comment by kif** [comment]
On Wed, 15 Apr 2020 12:29:24 -0700
ululi1970 <notifications@github.com> wrote:

> Compiling the serial version fails with the same problem 
> HDF5_DIR=/usr/lib/x86_64-linux-gnu/hdf5/serial/ pip3 install --no-binary=h5py h5py --user

If you aim at using mpi4py, you will need the MPI version on hdf5, not the serial one.

Did you install cython on your system to re-generate the c-files ?

Last remark: the HDFgroup advice NOT TO USE hdf5 1.10.x where x<3 as
they broke things, switched the size of ints in some structures which
makes files written with this version not readable with subsequent
versions of the library.

**Comment by Smankusors** [comment]
> this happens to me too
> 
> I think this is only happen when there's no compiled build/wheel available, and using Cython to build it.

Wait, I think I need to install cython as @kif mentioned. Now it's compiled successfully. But I wonder why it's not the requirement?

**Comment by ululi1970** [comment]
We have cython installed to regenerate the c-files, and the problem occurs
during the cythonization phase.
I just posted that to show that the problem occurred even if only the
serial version of the package was built.
We use hdf5 1.10.6.
We found out that using the latest sources pulled from the master branch on
github solves the compilation problem.
It appears that there was a commit pushed a couple of weeks ago that
addressed the issue (or so it seems).
Maybe it has not made its way yet to the tarball that pip downloads?

**Comment by takluyver** [comment]
Yes, any change in master won't be found until there's a new release, which is not imminent at the moment.

If you want to install the stable version, you'll need to identify what change has broken it (Cython 3?) and ensure you have a compatible version of that for building h5py.

**Comment by ululi1970** [comment]
I did some more digging into the problem.
Actually, I was wrong regarding Cython. We did not have Cython installed.
This means that when we install h5py with
pip3 install h5py
the installer downloads a temporary copy of Cython with version 3.0a1. Note
that this is the latest release that was released 4/12/2020 (five days
ago).
 With this version of Cython,
pip3 install h5py
fails as reported.
OTOH, dowloading the sources from the master branch of the repo, and using
pip3 install .
compiles fine, still using Cython 3.0a1.

Finally, if we install Cython with
pip3 install Cython
which installs Cython version 0.29.16
before running
pip3 install h5py
the process runs smoothly and h5py installs fine.

To summarize: compiling the files from master branch work with latest
version of Cython
compiling files from tarball requires previous version of Cython

Hope this helps you guys fixing the problem.

**Comment by ululi1970** [comment]
The problem is that if one tries to install h5py from binary using 
pip3 install --no-binary=h5py h5py 
and the user does not have cython installed, the install process will download the latest version of cython (30 alpha1), which is not compatible with h5py as delivered by the tarball. 
So in this case the solution is to install cython with
pip3 install cython
which installs version 0.29, and the install h5py. 

Alternatively, one can unstall the version from the code cloned from the github. HOWEVER, 
in this case make sure that you have numpy>=1.18.3, or else the library will compile, but will 
crash when loaded with the following error 

File "/nas/longleaf/home/XXX/.local/lib/python3.6/site-packages/h5py/__init__.py",
line 45, in <module>
    from ._conv import register_converters as _register_converters, \
  File "h5py/h5t.pxd", line 15, in init h5py._conv
  File "h5py/h5t.pyx", line 262, in init h5py.h5t
  File "h5py/h5t.pyx", line 257, in h5py.h5t._get_available_ftypes
AttributeError: 'numpy.dtype' object has no attribute 'typeobj'

 Apparently, the installer is satisfied with numpy version 1.7 or above, but that is clearly not enough.

**Comment by tacaswell** [comment]
@ululi1970 Can you make a new issue with the numpy minimum version issue?  That is definitely a bug.

**Comment by takluyver** [comment]
What you describe seems like it might also be a bug with pip - it shouldn't pick up an alpha release unless you use the `--pre` flag, and it sounds like it is picking that up when it installs Cython to build h5py.

**Comment by d-a-v** [comment]
Like @ululi1970 said above, installing cython first solved the problem for me.
I opened this issue and I will let it opened. Maintainers will decide to close it eventually.
Thanks everyone !

**Comment by aragilar** [comment]
I think the issue is setup_requires goes through setuptools, which grabs the latest release without worrying about whether something is a prerelease (though https://github.com/pypa/setuptools/issues/855 sees to suggest it's now using pip?). I'm not sure there much we can do in this case?

**Comment by takluyver** [comment]
Maybe using PEP 518 would avoid such problems, because pip would be responsible for installing the build dependencies before `setup.py` gets run. But IIRC we tried this at some point before and ran into other problems.

**Comment by takluyver** [comment]
AFAIK these issues with Cython 3 should be fixed in h5py 3.0. We've also made some changes to how our packaging works which hopefully makes things more reliable.
